### PR TITLE
add support for snapshoting namespaces

### PR DESCRIPTION
**Pull request by lexming**
Changelog:
* set `OceanStorRestClient` URL to the root of the API interface to be able to access other interfaces beyond `api/v2`
* add method to list active accounts
* add methods to list namespaces and get information about namespaces
* refactor method `list_filesystems()` to leverage `list_namespaces()`
* add method to list snapshots of a namespace
* add methods to create and delete snapshots of a namespace
* add unit tests for all new methods

### CODE REVIEW

File: lib/vsc/filesystem/oceanstor.py [hunk 1]
```diff
@@ -230,13 +228,17 @@ def request(self, method, url, body, headers, content_type=None):
             # Some queries generate a response with an int result
             # e.g. GET 'data_service/storagepool'
             exit_code = result
-            ec_msg = str(exit_code)
+            ec_msg_desc = ""
+            if "data" in response:
+                ec_msg_desc = response["data"]["errorMsg"]
```
Comment (wpoely86) [inline lib/vsc/filesystem/oceanstor.py]:
```suggestion
                ec_msg_desc = response["data"].get("errorMsg", "")
```

File: lib/vsc/filesystem/oceanstor.py [hunk 2]
```diff
@@ -295,46 +297,62 @@ def __init__(self, url, account, username, password):
...
+
         filter_json = [{"name": account_name}]
         filter_json = json.dumps(filter_json, separators=OCEANSTOR_JSON_SEP)
-        _, response = self.session.account.accounts.get(filter=filter_json)
-        ostor_account = response["data"][0]
+        _, response = self.session.api.v2.account.accounts.get(filter=filter_json)
+        try:
+            ostor_account = response["data"][0]
+        except IndexError:
+            errmsg = "OceanStor account not found: %s" % account_name
```
Comment (wpoely86) [inline lib/vsc/filesystem/oceanstor.py]: you can do python3 with f-string

File: lib/vsc/filesystem/oceanstor.py [hunk 3]
```diff
@@ -411,53 +429,118 @@ def select_storage_pools(self, sp_names, byid=False):
...
+                dbg_prefix = ""
+                # Query namespaces in all pools for this account
+                filter_json = [{"account_id": str(acc_id)}]
+                filter_json = json.dumps(filter_json, separators=OCEANSTOR_JSON_SEP)
+                _, response = self.session.api.v2.converged_service.namespaces.get(filter=filter_json)
+                # Save selection of attributes for this namespace
+                namespaces[acc_id] = {ns["name"]: {attr: ns[attr] for attr in ns_attrs} for ns in response["data"]}
+
+        # Update cache of namespaces with the selected accounts
+        self.oceanstor_namespaces.update(namespaces)
```
Comment (wpoely86) [inline lib/vsc/filesystem/oceanstor.py]: hmm, so you might update it again with the data you got from it in the first place? Can't we avoid that?

File: lib/vsc/filesystem/oceanstor.py [hunk 4]
```diff
@@ -411,53 +429,118 @@ def select_storage_pools(self, sp_names, byid=False):
...
+
+        # Filter on storage pools
         # Support special case 'all' for downstream compatibility
         if pool is None or pool == "all":
-            storage_pools = self.list_storage_pools(update=update)
+            storage_pools = self.list_storage_pools()
```
Comment (wpoely86) [inline lib/vsc/filesystem/oceanstor.py]: why not pass the flag?

File: lib/vsc/filesystem/oceanstor.py [hunk 5]
```diff
@@ -411,53 +429,118 @@ def select_storage_pools(self, sp_names, byid=False):
...
+        There is a single namespace in OceanStor, so there will be only namespace matching across all accounts
+
+        @type namespace: name of the namespace in OceanStor
+
+        @returns: dictionary with the namespace information
+
+        @raise OceanStorOperationError: if there is no namespace with the given name
+        """
+        oceanstor_namespaces = self.list_namespaces()
+        account = [ns for ns in oceanstor_namespaces if namespace in oceanstor_namespaces[ns]]
```
Comment (wpoely86) [inline lib/vsc/filesystem/oceanstor.py]: wait, and why this the account? you're looping over namespaces?

Comment (wpoely86) [reply lib/vsc/filesystem/oceanstor.py]: I wait, I think I understand but the name `ns` is poorly chosen ?

File: lib/vsc/filesystem/oceanstor.py [hunk 6]
```diff
@@ -1811,6 +1894,86 @@ def _get_grace_expiration(grace, usage, soft, hard):
 
         return expired
 
+    def list_namespace_snapshots(self, namespace):
```
Comment (wpoely86) [inline lib/vsc/filesystem/oceanstor.py:1911]: shouldn't we put an update flag here too?

File: lib/vsc/filesystem/oceanstor.py (hunk 2 above)
Comment (lexming) [reply lib/vsc/filesystem/oceanstor.py]: you want f-strings, you got f-strings https://github.com/vub-hpc/vsc-filesystem-oceanstor/pull/25/commits/1bee0ed7a140b58fa4f9414db633f6782091354c

File: lib/vsc/filesystem/oceanstor.py (hunk 3 above)
Comment (lexming) [reply lib/vsc/filesystem/oceanstor.py]: good point, that's indeed a waste. Fixed in https://github.com/vub-hpc/vsc-filesystem-oceanstor/pull/25/commits/f1039a4e15213f85acb9b0b3d8da27ded0bca06e

File: lib/vsc/filesystem/oceanstor.py (hunk 4 above)
Comment (lexming) [reply lib/vsc/filesystem/oceanstor.py]: that's an oversight, fixed in https://github.com/vub-hpc/vsc-filesystem-oceanstor/pull/25/commits/5ad66e6e08625c726a7722c16e7e570ac7ff4e22

File: lib/vsc/filesystem/oceanstor.py (hunk 5 above)
Comment (lexming) [reply lib/vsc/filesystem/oceanstor.py]: true, renamed to `acc` in https://github.com/vub-hpc/vsc-filesystem-oceanstor/pull/25/commits/b0a94c35c2253980b013c7a16059b6e0ea6e458b

File: lib/vsc/filesystem/oceanstor.py (hunk 6 above)
Comment (lexming) [reply lib/vsc/filesystem/oceanstor.py:1911]: there is no local cache for snapshots, not for namespaces or filesystems/dtrees. I think it is safer for snapshots to always query what exists in the storage.

**Comment by lexming** [comment]
Tests are good:
https://dev.azure.com/VUB-ICT/Directie%20ICT/_build/results?buildId=11706&view=results

**Review by wpoely86 — APPROVED** [review]

**Pull request merged**